### DISCUSSION TITLE
router.push doesn't work in production

### DISCUSSION THREAD

**Discussion by itzAymvn** [vercel/next.js#51782] [Help] [answered]
### Summary

This function is supposed to redirect the user to the `/dashboard` route if the sign in is successful, It works like that when `npm run dev` and everything perfect. But when I run the build (`npm run build` & `npm run start`) it just logs both "Before redirect" and "After redirect" but doesn't actually redirect.
```js
async function handleSubmit(event) {
    event.preventDefault();

    // reset message and error
    setMessage(null);
    setError(null);

    // get form data
    const form = event.target;
    const formData = new FormData(form);
    const email = formData.get("email");
    const password = formData.get("password");

    // validate
    if (!email || !password) {
        setError("Please fill in all fields");
        return;
    }

    // when everything is ok, change the content of the button to "Loading..."
    setButtonContent("Loading...");

    // send data to server
    try {
        const data = await signIn("credentials", {
            redirect: false,
            email,
            password,
        });

        if (data.error !== null) {
            setError(data.error);
            setButtonContent("Login");
            return;
        }
    } catch (error) {
        console.log(error);
        return;
    }

    console.log("Before redirect");
    router.push("/dashboard");
    console.log("After redirect");
}
```

**Comment by bartekczyz** [comment]
Experiencing exactly the same issue

**Reply by itzAymvn** [reply]
I made a fresh nextjs install and it works just perfect in the production build

**Comment by icyJoseph** [comment]
Is this using App Router? or Page Dir?

**Reply by itzAymvn** [reply]
App Router

**Reply by icyJoseph** [reply]
One more question, when a user arrives to dashboard, how do you send them to this `sign in` route?

**Reply by itzAymvn** [reply]
middleware.js in the root directory

**Reply by icyJoseph** [reply]
Aha, I think you fixed it? What a strange bug...

**Reply by itzAymvn** [reply]
No I didn't fix it, when the user put in correct data, it doesn't redirect them like it does when in dev mode

**Reply by icyJoseph** [reply]
The API route that handles signin, is that one of the new `route.ts` handlers?

**Reply by itzAymvn** [reply]
/api/auth/[...nextauth]/route.js

**Reply by itzAymvn** [reply]
Here is the repo: https://github.com/itzaymvn/aymvn.tech

**Reply by bartekczyz** [reply]
@itzAymvn did you have a chance to take a look at the repro repository? :)

**Comment by bartekczyz** [comment]
@itzAymvn I've created an issue for that: https://github.com/vercel/next.js/issues/52122

**Comment by bartekczyz** [comment, marked as answer]
@itzAymvn I don't know why exactly, but in my case it wouldn't redirect because at the time redirecting, router was not aware of user being successfully authenticated, so I had to refresh the state manually

```ts
router.refresh()
```

and then

```ts
router.replace('route-to-redirect')
```

So in your case it would be 


```ts
    console.log("Before redirect");
    router.refresh(); // dirty fix, but it works
    router.push("/dashboard");
    console.log("After redirect");
```

CC @icyJoseph

**Reply by itzAymvn** [reply]
Thank you for your solution @bartekczyz , currently the way I handled it is instead of redirecting directly, I set a "isLogged" state to true and then check inside a useEffect if its true then redirect. It's bad but worked for me

**Reply by samyarkd** [reply]
Next js docs:
> Refresh the current route. Making a new request to the server, re-fetching data requests, and re-rendering Server Components. The client will merge the updated React Server Component payload without losing unaffected client-side React (e.g. useState) or browser state (e.g. scroll position)

based on this description now it makes sense why we should use `router.refresh` because it will re-render server components without losing client-side state so it will know that the user is authenticated on the server, and it has access to the provided page

**Reply by saman2007** [reply]
Hello.
Im facing with this problem too. I have no idea how to fix it.
I tested your solutions but none of them worked. My situation is exactly like yours. I have a sign in form and when user sign in successfuly, i redirect the user to where the user came from. I have this behavior in some pages not all pages.
I moved the redirect functionality to a `useEffect`, but it didnt work. As im using `pages router`, i cant have access to `router.refresh`.
I didnt find anything about this problem on internet except this discussion. so do you have any idea about how can i can fix this problem?
thanks for helping.

@samyarkd @itzAymvn

**Reply by saman2007** [reply]
A i am using next js `pages router`, i added `await router.prefetch(path, path, {priority: true})` exactly before i call `router.push(path)` and it worked!

**Reply by devMiguelCarrero** [reply]
It's working with router.prefetch(). I had the same issue, thank you very much!

**Reply by nicolasheady** [reply]
Not even router.prefetch() is working for me. Neither does putting a flag inside useEffect. Neither does refresh() before push(), or replace().

**Reply by CaioMS2000** [reply]
I'm using Next 15.0.3, and I'm having same issue, router.push is triggering a GET request to the target route but not properly redirecting the user to the target route. And router.refresh is not working as well, so I'm using window.location.href = "/your-route" until I fix this.

**Reply by jamesopti** [reply]
Our issue is that middleware is running that does not get updated cookies that were set by a previous network request prior to calling `push`.

This has worked for us for 3+ years and suddenly broke.

**Reply by bartekczyz** [reply]
@jamesopti yeah it stopped working for me as well 😂 I followed @CaioMS2000's workaround with `globalThis.location.href = 'new_route'` and it worked.

**Reply by Manmit124** [reply]
@CaioMS2000 When I'm using  window.location.href ="/" this it goes into infinite loading 
next.js + supbase

**Reply by CaioMS2000** [reply]
> @CaioMS2000 When I'm using window.location.href ="/" this it goes into infinite loading next.js + supbase

I've had a problem before when redirecting users to the root URL. I'm not sure if it's exactly your case, but what happened to me was a "Too many redirects" error, if I remember correctly. I struggled quite a bit until I realized that a logic I had implemented in the middleware, which redirected users based on a certain condition, had a flaw in its logic. Every time it was redirected to the root, the middleware would redirect it again to the root, and that's what was causing my error. I don't know if your problem is similar to this.

**Reply by na-zi-ya** [reply]
> I'm using Next 15.0.3, and I'm having same issue, router.push is triggering a GET request to the target route but not properly redirecting the user to the target route. And router.refresh is not working as well, so I'm using window.location.href = "/your-route" until I fix this.

I tried using `window.location.href = '/'` to redirect to the home page, but after redirecting, I get an 'Unauthorized' error

**Comment by xiliumz** [comment]
I fixed the issue by removing the `router.prefetch('/dashboard')` line from my LoginForm component.

I believe the problem occurred because I attempted to prefetch the dashboard page before a user logged in. This caused an issue because I have middleware in place that redirects users to the root `/` page if they try to access the dashboard without a session cookie.

**Comment by Reynalroddy** [comment]
This issue is the most annoying ever..this is'nt my first time using next-auth..how tf will a router.push not work even after getting a console.log just before thee router.push...nothing seems to work..might just dump it for a custom auth..what a waste of time.damn

**Reply by Reynalroddy** [reply]
works on dev build..but for some reason production builds don't work.

**Comment by determinedcount** [comment]
Dealt with this same issue, the solution was similar to [this](https://github.com/vercel/next.js/discussions/51782#discussioncomment-8330050) from @xiliumz, but there was no prefetch(). If an unauth'd user clicked on a button that calls router.push('/edit'), middleware redirected them to /login. Then they would successfully auth and a router.push('/edit') on /login would run, but it did nothing. The solution was to point the original button at /login instead of /edit (middleware redirects already-auth'd users from /login to /edit anyway) and, for good measure, put a router.refresh() in front of /login's router.push().

**Comment by na-zi-ya** [comment]
I'm using Next.js 15, and router.push() is working fine in development but not in production after the user logs in. What could be the possible reasons, and how can I fix this..??

**Reply by stephenscript** [reply]
I faced this issue and tried the `router.refresh()` approach to no avail. Simply ditching the router and using `window.location.href = "/";` was what worked in my case.

You do lose some efficiencies with this switch, but until the root cause for this bug is fully understood I'll have to rely on this approach.

**Reply by argonathmos** [reply]
Using next 15.2.2.
My authentication mechanism is cookie-based. When a user logs in, the server sets an HTTP-only session cookie in the browser. Doing the router.refresh() then router.push("/here") did not work for me either.   

Had to use `window.location.href` as well. 

Note: I had to update my logout workflow to use window.location.href too.

**Comment by Vijji04** [comment]
here's something that fixed for me and my take on why it is failing for people especially using middleware.js.

IN LOCAL, middleware's position when placed inside src or at app/pages level would work without any issue and it fails to work when kept on the same level as package.json.

However ```opposite is true after build/production```, I changed my ```middleware.js to the root-level of project``` i.e same level as package.json and it worked, and I was ```able to use router.push``` without any issue.

**Reply by Coudercf** [reply]
Hello, 

Thank you for this work around. 

Indeed, when the `middleware.ts` file is located in the `src `folder, `router.push()` from the `next/navigation` library does not work correctly after a production build. 

I am using version **15.3.1** of next and **19.1.0** of react. 

So I moved the `middleware.ts` file to the root of my project next to the `package.json` file, and now everything works correctly.

**Reply by SergeySharipov** [reply]
Thanks a lot! Moving middleware.ts to the project root (next to package.json) fixed my issue in production too.

**Comment by supreme-gg-gg** [comment]
I ran into this problem when testing on Safari but once I switched to Chrome `router.push()` worked fine without any modifications. This seems to make the issue only more confusing, does anyone have similar experiences? Moving `middleware.ts` or using `window.location.href` both worked as workaround on Safari. Using **15.3.2** of next with bun, and both are in production build.

**Comment by ymabsout** [comment]
in my case i had a vercel.json file that had this "rewrites": [
    {
      "source": "/(.*)",
      "destination": "/"
    }
  ]
  This tells Vercel: "For ANY path /(.*)  → redirect to /"
  Removed it , and it worked.
